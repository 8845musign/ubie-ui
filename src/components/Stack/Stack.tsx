'use client';

import { clsx } from 'clsx';
import { isValidElement, cloneElement } from 'react';
import styles from './Stack.module.css';
import { CustomDataAttributeProps } from '../../types/attributes'; // 追加したインポート
import { Spacing, AlignItems, JustifyContent } from '../../types/style';
import { paddingVariables, marginVariables } from '../../utils/style';
import { HTMLTagname } from '../../utils/types';
import { Box } from '../Box/Box';
import type { PaddingProps, MarginProps } from '../../types/style';
import type { FC, ReactNode, ComponentType, ReactElement } from 'react';

type Props = {
  /**
   * レンダリングされるコンポーネントまたはHTML要素
   * @default div
   */
  as?: HTMLTagname | ReactElement<ComponentType<typeof Box>>;
  /**
   * 子要素の間隔
   * xxs=4px, xs=8px, sm=12px, md=16px, lg=24px, xl=40px, xxl=64px
   */
  spacing: Spacing;
  /**
   * 水平方向における子要素のレイアウトを定める。
   * @default stretch
   */
  alignItems?: AlignItems;
  /**
   * 垂直方向における子要素のレイアウトを定める。
   * @deprecated directionが削除されたため必要なくなりました
   */
  justifyContent?: JustifyContent;
  /**
   * CSSのクラス
   * @deprecated マージンなどをつけたい場合は<Box>を使ってください
   */
  className?: string;
  /**
   * 子要素
   */
  children: ReactNode;
} & MarginProps &
  PaddingProps &
  CustomDataAttributeProps;

/**
 * Stackコンポーネント
 * 一方向に一定のリズムで要素を積み上げるレイアウト
 */
export const Stack: FC<Props> = ({
  as: StackComponent = 'div',
  children,
  className,
  spacing,
  alignItems = 'stretch',
  justifyContent = 'flex-start',
  p,
  px,
  py,
  pt,
  pr,
  pb,
  pl,
  m,
  mx,
  my,
  mt,
  mr,
  mb,
  ml,
  ...otherProps
}) => {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const createElement = (props: any, children: ReactNode) => {
    if (isValidElement(StackComponent)) {
      return cloneElement(StackComponent, StackComponent.props, <div {...props}>{children}</div>);
    } else {
      return <StackComponent {...props}>{children}</StackComponent>;
    }
  };

  return createElement(
    {
      className: clsx(className, styles.stack),
      style: {
        alignItems: `${alignItems}`,
        justifyContent: `${justifyContent}`,
        gap: `var(--size-spacing-${spacing})`,
        ...paddingVariables({
          p,
          px,
          py,
          pt,
          pr,
          pb,
          pl,
        }),
        ...marginVariables({
          m,
          mx,
          my,
          mt,
          mr,
          mb,
          ml,
        }),
      },
      ...otherProps,
    },
    children,
  );
};
