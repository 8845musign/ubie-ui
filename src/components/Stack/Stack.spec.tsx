import { render, screen } from '@testing-library/react';
import { Stack } from './Stack';

describe('<Stack>', () => {
  it('has all the margins through m prop', () => {
    render(
      <Stack spacing="xs" m="xxs" data-testid="box">
        test
      </Stack>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-xxs)');
  });

  it('has a horizontal margins through mx prop', () => {
    render(
      <Stack spacing="xs" mx="xxs" data-testid="box">
        Test
      </Stack>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: 0');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-bottom: 0');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-xxs)');
  });

  it('has a vertical margins of through mx prop', () => {
    render(
      <Stack spacing="xs" my="xxs" data-testid="box">
        Test
      </Stack>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-right: 0');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-left: 0');
  });

  it('has all margins through individual margin props', () => {
    render(
      <Stack spacing="xs" mt="xxs" mr="xs" mb="sm" ml="md" data-testid="box">
        Test
      </Stack>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-xs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-sm)');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-md)');
  });

  it('gives priority to individual margin props', () => {
    render(
      <Stack spacing="xs" m="xxs" mx="xs" my="xs" mt="sm" mr="md" mb="lg" ml="xl" data-testid="box">
        Test
      </Stack>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-sm)');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-md)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-lg)');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-xl)');
  });
});
