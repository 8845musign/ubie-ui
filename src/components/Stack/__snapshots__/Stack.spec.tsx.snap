// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Stack> > Snapshot Test > changes rendered elements 1`] = `
<ul
  class="_stack_ccc237"
  data-testid="center"
  style="align-items: stretch; justify-content: flex-start; --gap: 0; --padding-top: var(--size-spacing-md); --padding-right: var(--size-spacing-md); --padding-bottom: var(--size-spacing-md); --padding-left: var(--size-spacing-md); --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --width: auto; --min-width: auto; --max-width: none;"
>
  <li>
    test
  </li>
  <li>
    test
  </li>
</ul>
`;

exports[`<Stack> > Snapshot Test > changes the rendered component 1`] = `
<div
  class="_stack_ccc237"
  data-testid="center"
  style="align-items: stretch; justify-content: flex-start; --gap: 0; --padding-top: var(--size-spacing-md); --padding-right: var(--size-spacing-md); --padding-bottom: var(--size-spacing-md); --padding-left: var(--size-spacing-md); --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --width: auto; --min-width: auto; --max-width: none;"
>
  <div>
    test
  </div>
  <div>
    test
  </div>
</div>
`;
